docs/foreign_code.md

```diff
@@ -0,0 +1,156 @@
+Verona will have to exist in a world with other languages but that should not compromise the security guarantees that we aim to provide.
+
+Core principles of interoperability
+-----------------------------------
+
+The Verona interoperability model will be built on the following principles:
+
+ - Foreign code should not imply unsafety.
+ - Interoperability should not compromise any security guarantees of the non-Verona language or library API.
+ - Most libraries are stateful, the correct granularity for interoperability is a library, not a function.
+
+The first of these relates to one of the core principles of Verona: there is no `unsafe` keyword.
+Verona targets security-critical workloads and it's very hard to make strong security claims about a language if any call to another language is able to compromise any and all of our safety guarantees.
+The Java JNI specification, for example, contains this disclaimer in the Design Overview chapter:
+
+> The programmer must not pass illegal pointers or arguments of the wrong type to JNI functions. Doing so could result in arbitrary consequences, including a corrupted system state or VM crash.
+
+The language makes an explicit declaration that any call to native code can undermine any of the invariants that the Java security model depends on. 
+This is a pragmatic decision to allow quick adoption but it means that no Java program can be considered safe until all native components have been removed.
+
+Modern C++ codebases work hard to provide memory safety guarantees at the API level.
+These guarantees are often lost in interop layers.
+For example, LLVM makes extensive use of typed smart pointers in the C++ APIs, but the Python bindings are written on top of the C API, which make type safety the caller's responsibility.
+Even if we can sandbox the resulting code, using a C++ library from Verona should not be more likely to introduce security vulnerabilities than using the same library from C++.
+The 2017 paper [*Finding and Preventing Bugs in JavaScript Bindings*](https://ieeexplore.ieee.org/document/7958598) showed that memory safety bugs in the C++ / JavaScript interop layer in Chrome are surprisingly common.
+Any time that a user has to hand-write code at a bindings layer to get memory safety right, we can assume that they will get it wrong and so we aim to automate as much as possible.
+
+The desire for sandboxing also leads to the conclusion that a 'foreign function interface' is exposing the wrong level of abstraction.
+What sandbox should a function run in?
+Our prior [CHERI-JNI](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf) work showed that treating libraries (in the loose sense, including any dependent libraries) as the unit of foreign code provided an abstraction that fitted well with sandboxing and allowed the user to determine the degree of state sharing between foreign code invocations.
```

Comment:
Perhaps remove "Our".

Reply:
I want to call out that we're building on prior experience with designing safe interop systems.  Can you suggest a way of doing that?

Reply:
Does it mater that it is our experience.   I guess I have written too many double blind reviewed papers now.  Feel free to ignore.